javalin/src/main/java/io/javalin/http/Context.kt

```diff
@@ -26,132 +23,91 @@ import java.nio.charset.Charset
-     */
-    fun endpointHandlerPath() = if (handlerType != HandlerType.BEFORE) {
-        endpointHandlerPath
-    } else {
-        throw IllegalStateException("Cannot access the endpoint handler path in a 'BEFORE' handler")
-    }
+    fun <T> appAttribute(key: String): T
+    /** Gets the handler type of the current handler */
+    fun handlerType(): HandlerType
+    /** * Gets the path that was used to match request (also includes before/after paths) */
+    fun matchedPath(): String
+    /** Gets the endpoint path that was used to match request (null in before, available in endpoint/after) */
+    fun endpointHandlerPath(): String
 
     ///////////////////////////////////////////////////////////////
     // Request-ish methods
     ///////////////////////////////////////////////////////////////
 
-    /** Gets the request body as a [String]. */
-    fun body(): String = bodyAsBytes().toString(Charset.forName(characterEncoding))
+    /** Gets the request content length. */
+    fun contentLength(): Int = req.contentLength
+    /** Gets the request content type, or null. */
+    fun contentType(): String? = req.contentType
+    /** Gets the request method. */
+    fun method(): HandlerType
+    /** Gets the request path. */
+    fun path(): String = req.requestURI
+    /** Gets the request port. */
+    fun port(): Int = req.serverPort
+    /** Gets the request protocol. */
+    fun protocol(): String = req.protocol
+    /** Gets the request context path. */
+    fun contextPath(): String = req.contextPath
+    /** Gets the request user agent, or null. */
+    fun userAgent(): String? = req.getHeader(Header.USER_AGENT)
+    /** Try to obtain request encoding from [Header.CONTENT_TYPE] header */
+    fun characterEncoding(): String?
+
+    /** Gets the request url. */
+    fun url(): String = contextResolver().url.invoke(this)
+    /** Gets the full request url, including query string (if present) */
+    fun fullUrl(): String = contextResolver().fullUrl.invoke(this)
+    /** Gets the request scheme. */
+    fun scheme(): String = contextResolver().scheme.invoke(this)
+    /** Gets the request host, or null. */
+    fun host(): String? = contextResolver().host.invoke(this)
+    /** Gets the request ip. */
+    fun ip(): String = contextResolver().ip.invoke(this)
 
-    /** Gets the request body as a [ByteArray].
+    /** Gets the request body as a [String]. */
+    fun body(): String = bodyAsBytes().toString(Charset.forName(characterEncoding() ?: "UTF-8"))
+    /**
+     * Gets the request body as a [ByteArray].
      * Calling this method returns the body as a [ByteArray]. If [io.javalin.JavalinConfig.maxRequestSize]
      * is set and body is bigger than its value, a [io.javalin.http.HttpResponseException] is throw,
      * with status 413 CONTENT_TOO_LARGE.
      */
-    fun bodyAsBytes(): ByteArray = body
-
+    fun bodyAsBytes(): ByteArray
```

Comment:
Why can't this have a default implementation?

Reply:
Because it's bounded with field and all those cached properties are in ServletContext. Also, I didn't want to implement the same methods on both sides with different impl:

https://github.com/javalin/javalin/blob/9a21d54897d16c2393b9b8e4bd74c828d39e3ba8/javalin/src/main/java/io/javalin/http/ServletContext.kt#L62